File: go+reevit/webhooks.go
package main

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"io"
	"log"
	"net/http"
	"os"
	"strings"
)

// WebhookPayload represents incoming webhook event
type WebhookPayload struct {
	ID        string          `json:"id"`
	Type      string          `json:"type"`
	OrgID     string          `json:"org_id"`
	CreatedAt string          `json:"created_at"`
	Data      json.RawMessage `json:"data,omitempty"`
	Message   string          `json:"message,omitempty"`
}

// PaymentEventData represents payment event data
type PaymentEventData struct {
	ID         string            `json:"id"`
	Status     string            `json:"status"`
	Amount     int64             `json:"amount"`
	Currency   string            `json:"currency"`
	Provider   string            `json:"provider"`
	CustomerID string            `json:"customer_id,omitempty"`
	Metadata   map[string]string `json:"metadata,omitempty"`
}

// VerifySignature verifies webhook signature (HMAC-SHA256)
func VerifySignature(payload []byte, signature, secret string) bool {
	if !strings.HasPrefix(signature, "sha256=") {
		return false
	}

	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write(payload)
	expected := hex.EncodeToString(mac.Sum(nil))
	received := signature[7:] // Remove "sha256=" prefix

	return hmac.Equal([]byte(received), []byte(expected))
}

// HandleWebhook processes incoming webhooks from Reevit
func (s *Server) HandleWebhook(w http.ResponseWriter, r *http.Request) {
	// Read raw body
	body, err := io.ReadAll(r.Body)
	if err != nil {
		http.Error(w, `{"error": "Failed to read body"}`, http.StatusBadRequest)
		return
	}
	defer r.Body.Close()

	// Verify signature (required in production)
	signature := r.Header.Get("X-Reevit-Signature")
	secret := os.Getenv("REEVIT_WEBHOOK_SECRET")

	if secret != "" && !VerifySignature(body, signature, secret) {
		log.Println("[Webhook] Invalid signature")
		http.Error(w, `{"error": "Invalid signature"}`, http.StatusUnauthorized)
		return
	}

	// Parse event
	var event WebhookPayload
	if err := json.Unmarshal(body, &event); err != nil {
		http.Error(w, `{"error": "Invalid JSON"}`, http.StatusBadRequest)
		return
	}

	log.Printf("[Webhook] Received: %s (%s)", event.Type, event.ID)

	// Handle different event types
	switch event.Type {
	case "reevit.webhook.test":
		log.Printf("[Webhook] Test event: %s", event.Message)

	case "payment.succeeded":
		var data PaymentEventData
		if err := json.Unmarshal(event.Data, &data); err == nil {
			handlePaymentSucceeded(data)
		}

	case "payment.failed":
		var data PaymentEventData
		if err := json.Unmarshal(event.Data, &data); err == nil {
			handlePaymentFailed(data)
		}

	case "payment.refunded":
		var data PaymentEventData
		if err := json.Unmarshal(event.Data, &data); err == nil {
			handlePaymentRefunded(data)
		}

	default:
		log.Printf("[Webhook] Unhandled event: %s", event.Type)
	}

	// Acknowledge receipt
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]bool{"received": true})
}

func handlePaymentSucceeded(data PaymentEventData) {
	orderID := data.Metadata["order_id"]
	log.Printf("[Webhook] ✅ Payment succeeded: %s for order %s", data.ID, orderID)

	// TODO: Implement your business logic
	// - Update order status to "paid"
	// - Send confirmation email
	// - Trigger fulfillment
}

func handlePaymentFailed(data PaymentEventData) {
	log.Printf("[Webhook] ❌ Payment failed: %s", data.ID)

	// TODO: Implement your business logic
	// - Update order status
	// - Notify customer
}

func handlePaymentRefunded(data PaymentEventData) {
	orderID := data.Metadata["order_id"]
	log.Printf("[Webhook] 💰 Payment refunded: %s for order %s", data.ID, orderID)

	// TODO: Implement your business logic
	// - Update order status
	// - Restore inventory
}

File: python+reevit/webhooks.py
import os
import hmac
import hashlib
import logging
from fastapi import APIRouter, Request, HTTPException

from models import WebhookEvent, PaymentEventData

logger = logging.getLogger(__name__)

router = APIRouter()


def verify_signature(payload: bytes, signature: str, secret: str) -> bool:
    """Verify webhook signature using HMAC-SHA256."""
    if not signature.startswith("sha256="):
        return False
    
    expected = hmac.new(
        secret.encode("utf-8"),
        payload,
        hashlib.sha256
    ).hexdigest()
    
    received = signature[7:]  # Remove "sha256=" prefix
    return hmac.compare_digest(received, expected)


@router.post("/webhooks/reevit")
async def handle_webhook(request: Request):
    """Handle incoming webhooks from Reevit."""
    payload = await request.body()
    signature = request.headers.get("X-Reevit-Signature", "")
    secret = os.getenv("REEVIT_WEBHOOK_SECRET", "")
    
    # Verify signature (required in production)
    if secret and not verify_signature(payload, signature, secret):
        logger.warning("[Webhook] Invalid signature")
        raise HTTPException(status_code=401, detail="Invalid signature")
    
    # Parse event
    event_data = await request.json()
    event = WebhookEvent(**event_data)
    
    logger.info(f"[Webhook] Received: {event.type} ({event.id})")
    
    # Handle different event types
    if event.type == "reevit.webhook.test":
        logger.info(f"[Webhook] Test event: {event.message}")
    
    elif event.type == "payment.succeeded":
        if event.data:
            data = PaymentEventData(**event.data)
            handle_payment_succeeded(data)
    
    elif event.type == "payment.failed":
        if event.data:
            data = PaymentEventData(**event.data)
            handle_payment_failed(data)
    
    elif event.type == "payment.refunded":
        if event.data:
            data = PaymentEventData(**event.data)
            handle_payment_refunded(data)
    
    else:
        logger.info(f"[Webhook] Unhandled event: {event.type}")
    
    return {"received": True}


def handle_payment_succeeded(data: PaymentEventData):
    """Handle successful payment."""
    order_id = data.metadata.get("order_id") if data.metadata else None
    logger.info(f"[Webhook] ✅ Payment succeeded: {data.id} for order {order_id}")
    
    # TODO: Implement your business logic
    # - Update order status to "paid"
    # - Send confirmation email
    # - Trigger fulfillment


def handle_payment_failed(data: PaymentEventData):
    """Handle failed payment."""
    logger.info(f"[Webhook] ❌ Payment failed: {data.id}")
    
    # TODO: Implement your business logic
    # - Update order status
    # - Notify customer


def handle_payment_refunded(data: PaymentEventData):
    """Handle refunded payment."""
    order_id = data.metadata.get("order_id") if data.metadata else None
    logger.info(f"[Webhook] 💰 Payment refunded: {data.id} for order {order_id}")
    
    # TODO: Implement your business logic
    # - Update order status
    # - Restore inventory
